expanding on significant terms (#11178)
opensearch-project/documentation-website: _aggregations/bucket/significant-terms.md

- Foreground set: the documents matched by your query.
- Background set: by default, all documents in the target indexes. You can narrow it with `background_filter`.

Each result bucket includes:

- `key`: the term value.
- `doc_count`: the number of foreground docs containing the term.
- `bg_count`: the number of background docs containing the term.
- `score`: how strongly the term stands out in the foreground relative to the background, see [Heuristics and scoring](#heuristics-and-scoring) for further details.

If the aggregation returns no buckets, you likely didn’t filter the foreground, for example using `match_all`, or the foreground has the same distribution of terms as the background.
{: .note}

## Basic example: Find out what's distinctive about high‑value returns

You can use the following query to ask, among orders that were *returned and cost over 500*, which `payment_method` values are unusually common compared to the whole index?

```json

> GET retail_orders/_search

Reviewer: I suggest to also provide example indexing commands and the output of this search request, so user can easily follow.

# Significant terms aggregations

> `significant_terms` helps you surface terms that are unusually frequent in a subset of documents (foreground set) compared to a broader reference set (background set). It’s the right choice when a plain `terms` aggregation shows you the *most common* values, but you want the *most over‑represented* values.

Reviewer 2: ```suggestion
The `significant_terms` aggregation identifies terms that occur unusually frequently in a subset of documents (foreground set) compared to a broader reference set (background set). Use this aggregation to retrieve the *most over‑represented* values, for which a plain `terms` aggregation that shows you the *most common* values is not sufficient.
```

- Foreground set: the documents matched by your query.
- Background set: by default, all documents in the target indexes. You can narrow it with `background_filter`.

Each result bucket includes:

- `key`: the term value.
- `doc_count`: the number of foreground docs containing the term.
- `bg_count`: the number of background docs containing the term.
- `score`: how strongly the term stands out in the foreground relative to the background, see [Heuristics and scoring](#heuristics-and-scoring) for further details.

If the aggregation returns no buckets, you likely didn’t filter the foreground, for example using `match_all`, or the foreground has the same distribution of terms as the background.
{: .note}

## Basic example: Find out what's distinctive about high‑value returns

Create index mapping::

```json
PUT retail_orders
{
  "mappings": {
    "properties": {
      "status":         { "type": "keyword" },
      "order_total":    { "type": "double" },
      "payment_method": { "type": "keyword" }
    }
  }
}
```
{% include copy-curl.html %}

> Index sample documents:

Reviewer 2: ```suggestion
Ingest sample documents into the index:
```

The following request segments by tiles and queries which `issue.keyword` is unusually frequent in those tiles:

```json
GET field_ops/_search
{
  "size": 0,
  "aggs": {
    "tiles": {
      "geotile_grid": { "field": "site.location", "precision": 6 },
      "aggs": {
        "odd_issues": {
          "significant_terms": { "field": "issue.keyword" }
        }
      }
    }
  }
}
```
{% include copy-curl.html %}

## Focus the background with `background_filter`

By default, the background is the entire index. Sometimes you want a **narrower reference set**.

> ### Example: Toronto verses the rest of Canada

Reviewer 2: ```suggestion
### Example: Compare Toronto to the rest of Canada
```

> 

Reviewer 2: Add a descriptive intro sentence for the request

## Heuristics and scoring

The `score` ranks terms by how much their foreground frequency departs from the background frequency. It has no units and is meaningful only for comparison within the same request and heuristic.

You can choose one heuristic per request by adding its object under `significant_terms`.

### JLH (balanced absolute × relative lift)

Good general‑purpose choice. Favors terms that increase both *absolutely* and *relatively*.

```json
"significant_terms": {
  "field": "payment_method.keyword",
  "jlh": {}
}
```

#### JLH scoring

JLH score is calculated as follows:

`fg_pct = doc_count / foreground_total` and `bg_pct = bg_count / background_total`. JLH ≈ `(fg_pct − bg_pct) * (fg_pct / bg_pct)`. A term whose share rises a little but from a much bigger baseline will rank higher than one with the same absolute rise from a tiny background share.

#### Score example calculation using JLH

> If your foreground (high‑value returns) has `2,000` orders and the background (all orders) has `120,000`. One bucket reports:

Reviewer 2: ```suggestion
Suppose that your foreground set (high‑value returns) contains `2,000` orders, and the background set (all orders) contains `120,000` orders. Consider a single term in the `significant_terms` aggregation, which has the following counts:
```

  "size": 0,
  "query": {
    "bool": {
      "filter": [
        { "term":  { "status": "RETURNED" } },
        { "range": { "order_total": { "gte": 500 } } }
      ]
    }
  },
  "aggs": {
    "payment_signals": {
      "significant_terms": {
        "field": "payment_method"
      }
    }
  }
}
```
{% include copy-curl.html %}

The returned aggregation shows that among the five high-value returns, `gift_card` appears 3 times (60%), compared to 3 out of 12 times in the entire index (25%). As a result, it is flagged as the most overrepresented payment method:

```json
{
  ...
  "hits": {
    "total": {
      "value": 5,
      "relation": "eq"
    },
    "max_score": null,
    "hits": []
  },
  "aggregations": {
    "payment_signals": {
      "doc_count": 5,
      "bg_count": 12,
      "buckets": [
        {
          "key": "gift_card",
          "doc_count": 3,
          "score": 0.84,
          "bg_count": 3
        }
      ]
    }
  }
}
```

## Multi‑set analysis

You can determine the unusual values for each category by first grouping documents into buckets and then running a `significant_terms` aggregation within each bucket.

### Example: Unusual `cancel_reason` per region

> The following example groups by region with a terms aggregation, and within each bucket runs `significant_terms` to reveal cancellation reasons disproportionately common in that region:

Reviewer 3: ```suggestion
The following example groups by region with a terms aggregation and within each bucket runs `significant_terms` to identify cancellation reasons disproportionately common in that region:
```

```json
"significant_terms": {
  "field": "payment_method.keyword",
  "jlh": {}
}
```

#### JLH scoring

JLH score is calculated as follows:

`fg_pct = doc_count / foreground_total` and `bg_pct = bg_count / background_total`. JLH ≈ `(fg_pct − bg_pct) * (fg_pct / bg_pct)`. 

A term whose frequency increases slightly from a large baseline will score higher than a term with the same absolute increase from a very small background share.

#### Score example calculation using JLH

Suppose that your foreground set (high‑value returns) contains `2,000` orders, and the background set (all orders) contains `120,000` orders. Consider a single term in the `significant_terms` aggregation, which has the following counts:

- `doc_count = 160`
- `bg_count = 3,200`

Percentages of documents containing the term are calculated as follows:

- `fg_pct = 160 / 2000 = 0.08`
- `bg_pct = 3200 / 120000 ≈ 0.026666…`

JLH ≈ `(0.08 − 0.026666…) * (0.08 / 0.026666…) ≈ 0.053333… * 3 ≈ 0.16`

> This positive score means that the searched term is notably more prevalent in high‑value returns than overall. Scores are relative: use them to rank terms, not as absolute probabilities.

Reviewer 3: ```suggestion
This positive score means that the searched term is notably more prevalent in high‑value returns than it is overall. Scores are relative: use them to rank terms, not as absolute probabilities.
```

### Mutual information

Mutual information (MI) prefers frequent terms and identifies popular but still distinctive terms. Set `include_negatives: false` to ignore terms that are less common in the foreground than the background. If your background is not a superset of the foreground, set `background_is_superset: false`:

```json
"significant_terms": {
  "field": "product.keyword",
  "mutual_information": {
    "include_negatives": false,
    "background_is_superset": true
  }
}
```

### Chi‑square

> Chi-square is a statistical test that measures how much the observed frequency of a term in a subset (foreground) deviates from the expected frequency based on a reference set (background). Similarly to [MI](#mutual-information),  chi-square supports `include_negatives` and `background_is_superset`.

Reviewer 3: ```suggestion
Chi-square is a statistical test that measures how much the observed frequency of a term in a subset (foreground) deviates from the expected frequency based on a reference set (background). Similarly to [MI](#mutual-information), chi-square supports `include_negatives` and `background_is_superset`:
```

```json
"significant_terms": {
  "field": "error.keyword",
  "chi_square": { "include_negatives": false }
}
```

### Google Normalized Distance

Google Normalized Distance (GND) favors strong co‑occurrence. It is useful for synonym discovery or items that tend to appear together.

```json
"significant_terms": {
  "field": "tag.keyword",
  "gnd": {}
}
```

### Percentage

Percentage sorts terms by the `doc_count`/`bg_count` ratio and identifies the number of foreground hits a term has relative to its background hits. It doesn’t account for the overall sizes of the two sets, so very rare terms can dominate.

```json
"significant_terms": {
  "field": "sku.keyword",
  "percentage": {}
}
```

### Scripted heuristic

To provide a custom heuristic formula, use the following variables:

- `_subset_freq`: The number of documents containing the term in the foreground set.
- `_superset_freq`: The number of documents containing the term in the background set.
- `_subset_size`: The total number of documents in the foreground set.
- `_superset_size`: The total number of documents in the background set.

> The following request runs a `significant_terms` aggregation on `field.keyword` using a custom script heuristic to score terms based on their frequency in the foreground relative to the background:

Reviewer 3: This line is duplicated.
